// nativescript-toasty/toasty.android.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("tns-core-modules/color");
var platform_1 = require("tns-core-modules/platform");
var utils_1 = require("tns-core-modules/utils/utils");
var toast_common_1 = require("./toast.common");
var types_1 = require("tns-core-modules/utils/types");
var application_1 = require("tns-core-modules/application");
__export(require("./toast.common"));
var Toasty = (function () {
    function Toasty(opts) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
        this._screenHeight = platform_1.screen.mainScreen.heightPixels;
        this._screenWidth = platform_1.screen.mainScreen.widthPixels;
        this._text = (_a = opts) === null || _a === void 0 ? void 0 : _a.text;
        this._duration = (_c = (_b = opts) === null || _b === void 0 ? void 0 : _b.duration, (_c !== null && _c !== void 0 ? _c : toast_common_1.ToastDuration.SHORT));
        this._position = (_e = (_d = opts) === null || _d === void 0 ? void 0 : _d.position, (_e !== null && _e !== void 0 ? _e : toast_common_1.ToastPosition.BOTTOM));
        this._textColor = (_g = (_f = opts) === null || _f === void 0 ? void 0 : _f.textColor, (_g !== null && _g !== void 0 ? _g : 'white'));
        this._backgroundColor = (_j = (_h = opts) === null || _h === void 0 ? void 0 : _h.backgroundColor, (_j !== null && _j !== void 0 ? _j : 'black'));
        this._androidOpts = (_l = (_k = opts) === null || _k === void 0 ? void 0 : _k.android, (_l !== null && _l !== void 0 ? _l : {}));
        this._x = (_m = opts) === null || _m === void 0 ? void 0 : _m.xAxisOffset;
        this._y = (_o = opts) === null || _o === void 0 ? void 0 : _o.yAxisOffset;
        this._tapToDismiss = (_q = (_p = opts) === null || _p === void 0 ? void 0 : _p.tapToDismiss, (_q !== null && _q !== void 0 ? _q : false));
        this._anchorView = (_r = opts) === null || _r === void 0 ? void 0 : _r.anchorView;
        this._toast = android.widget.Toast.makeText(utils_1.ad.getApplicationContext(), this._text, android.widget.Toast.LENGTH_SHORT);
        var ref = new WeakRef(this);
        this._toast.getView().setOnTouchListener(new android.view.View.OnTouchListener({
            onTouch: function (param0, param1) {
                var _a, _b, _c;
                if ((_a = ref.get()) === null || _a === void 0 ? void 0 : _a._tapToDismiss) {
                    (_c = (_b = ref.get()) === null || _b === void 0 ? void 0 : _b._toast) === null || _c === void 0 ? void 0 : _c.cancel();
                }
                return false;
            }
        }));
        this.setToastDuration(this._duration)
            .setToastPosition(this._position)
            .setTextColor(this._textColor)
            .setBackgroundColor(this._backgroundColor);
        return this;
    }
    Object.defineProperty(Toasty.prototype, "position", {
        get: function () {
            return this._position;
        },
        set: function (value) {
            if (value) {
                this._position = value;
                this.setToastPosition(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "duration", {
        get: function () {
            return this._duration;
        },
        set: function (value) {
            if (value) {
                this._duration = value;
                this.setToastDuration(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "textColor", {
        get: function () {
            return this._textColor;
        },
        set: function (value) {
            if (value) {
                this._textColor = value;
                this.setTextColor(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "backgroundColor", {
        get: function () {
            return this._backgroundColor;
        },
        set: function (value) {
            if (value) {
                this._backgroundColor = value;
                this.setBackgroundColor(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "xAxisOffset", {
        get: function () {
            return this._x;
        },
        set: function (value) {
            this._x = value;
            this._updateToastPosition();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "yAxisOffset", {
        get: function () {
            return this._y;
        },
        set: function (value) {
            this._y = value;
            this._updateToastPosition();
        },
        enumerable: true,
        configurable: true
    });
    Toasty.isLength = function (value) {
        return (value && !types_1.isNullOrUndefined(value.unit) && !types_1.isNullOrUndefined(value.value));
    };
    Object.defineProperty(Toasty.prototype, "yPixels", {
        get: function () {
            var y;
            if (typeof this._y === 'number') {
                y = this._y * platform_1.screen.mainScreen.scale;
            }
            if (Toasty.isLength(this._y)) {
                if (this._y.unit === 'px') {
                    y = this._y.value;
                }
                if (this._y.unit === 'dip') {
                    y = this._y.value * platform_1.screen.mainScreen.scale;
                }
            }
            if (typeof this._y === 'string') {
                if (this._y.includes('px')) {
                    y = parseInt(this._y.replace('px', ''), 0);
                }
                if (this._y.includes('dip')) {
                    y = parseInt(this._y.replace('dip', ''), 0) * platform_1.screen.mainScreen.scale;
                }
            }
            return y;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "xPixels", {
        get: function () {
            var x;
            if (typeof this._x === 'number') {
                x = this._x * platform_1.screen.mainScreen.scale;
            }
            if (Toasty.isLength(this._x)) {
                if (this._x.unit === 'px') {
                    x = this._x.value;
                }
                if (this._x.unit === 'dip') {
                    x = this._x.value * platform_1.screen.mainScreen.scale;
                }
            }
            if (typeof this._x === 'string') {
                if (this._x.includes('px')) {
                    x = parseInt(this._x.replace('px', ''), 0);
                }
                if (this._x.includes('dip')) {
                    x = parseInt(this._x.replace('dip', ''), 0) * platform_1.screen.mainScreen.scale;
                }
            }
            return x;
        },
        enumerable: true,
        configurable: true
    });
    Toasty.prototype.show = function () {
        var _a;
        if (!this._text) {
            throw new Error('Toasty Text is not set.');
        }
        else {
            (_a = this._toast) === null || _a === void 0 ? void 0 : _a.show();
        }
    };
    Toasty.prototype._measureToast = function () {
        var _a, _b, _c;
        var window = (_a = (application_1.android.foregroundActivity || application_1.android.startActivity)) === null || _a === void 0 ? void 0 : _a.getWindow();
        var metrics = new android.util.DisplayMetrics();
        (_b = window) === null || _b === void 0 ? void 0 : _b.getWindowManager().getDefaultDisplay().getMetrics(metrics);
        var MeasureSpec = android.view.View.MeasureSpec;
        var widthMeasureSpec = MeasureSpec.makeMeasureSpec(metrics.widthPixels, MeasureSpec.UNSPECIFIED);
        var heightMeasureSpec = MeasureSpec.makeMeasureSpec(metrics.heightPixels, MeasureSpec.UNSPECIFIED);
        (_c = this._toast) === null || _c === void 0 ? void 0 : _c.getView().measure(widthMeasureSpec, heightMeasureSpec);
    };
    Object.defineProperty(Toasty.prototype, "width", {
        get: function () {
            var _a, _b;
            this._measureToast();
            return (_b = ((_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView().getMeasuredWidth()) / platform_1.screen.mainScreen.scale, (_b !== null && _b !== void 0 ? _b : 0));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Toasty.prototype, "height", {
        get: function () {
            var _a, _b;
            this._measureToast();
            return (_b = ((_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView().getMeasuredHeight()) / platform_1.screen.mainScreen.scale, (_b !== null && _b !== void 0 ? _b : 0));
        },
        enumerable: true,
        configurable: true
    });
    Toasty.prototype.cancel = function () {
        var _a;
        (_a = this._toast) === null || _a === void 0 ? void 0 : _a.cancel();
    };
    Toasty.prototype.setTextColor = function (value) {
        var _a, _b, _c, _d;
        if (value) {
            this._textColor = value;
            var view = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView();
            var text = (_b = view) === null || _b === void 0 ? void 0 : _b.findViewById(android.R.id.message);
            if (typeof value === 'string') {
                var nativeColor = new color_1.Color(value).android;
                (_c = text) === null || _c === void 0 ? void 0 : _c.setTextColor(nativeColor);
            }
            else {
                (_d = text) === null || _d === void 0 ? void 0 : _d.setTextColor(value.android);
            }
        }
        return this;
    };
    Toasty.prototype.setBackgroundColor = function (value) {
        var _a, _b, _c;
        if (value) {
            this._backgroundColor = value;
            var view = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView();
            if (typeof value === 'string') {
                var nativeColor = new color_1.Color(value).android;
                (_b = view) === null || _b === void 0 ? void 0 : _b.setBackgroundColor(nativeColor);
            }
            else {
                (_c = view) === null || _c === void 0 ? void 0 : _c.setBackgroundColor(value.android);
            }
        }
        return this;
    };
    Toasty.prototype.setToastDuration = function (value) {
        var _a, _b, _c;
        switch (value) {
            case toast_common_1.ToastDuration.SHORT:
                (_a = this._toast) === null || _a === void 0 ? void 0 : _a.setDuration(android.widget.Toast.LENGTH_SHORT);
                break;
            case toast_common_1.ToastDuration.LONG:
                (_b = this._toast) === null || _b === void 0 ? void 0 : _b.setDuration(android.widget.Toast.LENGTH_LONG);
                break;
            default:
                (_c = this._toast) === null || _c === void 0 ? void 0 : _c.setDuration(android.widget.Toast.LENGTH_SHORT);
                break;
        }
        return this;
    };
    Toasty.prototype.setToastPosition = function (value) {
        this._position = value;
        this._updateToastPosition();
        return this;
    };
    Toasty.prototype._updateToastPosition = function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        var yOffset = (_a = this.yPixels, (_a !== null && _a !== void 0 ? _a : 0));
        var xOffset = (_b = this.xPixels, (_b !== null && _b !== void 0 ? _b : 0));
        switch (this._position) {
            case toast_common_1.ToastPosition.TOP:
                (_c = this._toast) === null || _c === void 0 ? void 0 : _c.setGravity(android.view.Gravity.TOP, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.TOP_LEFT:
                (_d = this._toast) === null || _d === void 0 ? void 0 : _d.setGravity(android.view.Gravity.TOP | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.TOP_RIGHT:
                (_e = this._toast) === null || _e === void 0 ? void 0 : _e.setGravity(android.view.Gravity.TOP | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.CENTER:
                (_f = this._toast) === null || _f === void 0 ? void 0 : _f.setGravity(android.view.Gravity.CENTER, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.CENTER_LEFT:
                (_g = this._toast) === null || _g === void 0 ? void 0 : _g.setGravity(android.view.Gravity.CENTER | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.CENTER_RIGHT:
                (_h = this._toast) === null || _h === void 0 ? void 0 : _h.setGravity(android.view.Gravity.CENTER | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.BOTTOM:
                (_j = this._toast) === null || _j === void 0 ? void 0 : _j.setGravity(android.view.Gravity.BOTTOM, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.BOTTOM_LEFT:
                (_k = this._toast) === null || _k === void 0 ? void 0 : _k.setGravity(android.view.Gravity.BOTTOM | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case toast_common_1.ToastPosition.BOTTOM_RIGHT:
                (_l = this._toast) === null || _l === void 0 ? void 0 : _l.setGravity(android.view.Gravity.BOTTOM | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            default:
                break;
        }
    };
    return Toasty;
}());
exports.Toasty = Toasty;
